Content.Shared/Whistle/WhistleSystem.cs

```diff
@@ -0,0 +1,71 @@
+using Content.Shared.Coordinates;
+using Content.Shared.Humanoid;
+using Content.Shared.Interaction.Events;
+using Content.Shared.Stealth.Components;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Whistle;
+
+public sealed class WhistleSystem : EntitySystem
+{
+    [Dependency] private readonly EntityLookupSystem _entityLookup = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<WhistleComponent, UseInHandEvent>(OnUseInHand);
+    }
+
+    private void ExclamateTarget(EntityUid target, WhistleComponent component)
+    {
+        SpawnAttachedTo(component.Effect, target.ToCoordinates());
+    }
+
+    public void OnUseInHand(EntityUid uid, WhistleComponent component, UseInHandEvent args)
+    {
+        if (!_timing.IsFirstTimePredicted)
+            return;
+
+        TryMakeLoudWhistle(uid, args.User, component);
+        args.Handled = true;
+    }
+
+    public bool TryMakeLoudWhistle(EntityUid uid, EntityUid owner, WhistleComponent? component = null)
+    {
+        if (!Resolve(uid, ref component, false) || component.Distance <= 0)
+            return false;
+
+        MakeLoudWhistle(uid, owner, component);
+        return true;
+    }
+
+    private void MakeLoudWhistle(EntityUid uid, EntityUid owner, WhistleComponent component)
+    {
+        StealthComponent? stealth = null;
+
+        foreach (var iterator in
+            _entityLookup.GetEntitiesInRange<HumanoidAppearanceComponent>(_transform.GetMapCoordinates(uid), component.Distance))
+        {
+            //Avoid pinging invisible entities
+            if (Resolve(iterator, ref stealth, false))
+            {
+                if (stealth.Enabled)
+                {
+                    stealth = null;
+                    continue;
+                }
+
+                stealth = null;
+            }
+
+            //We don't want to ping user of whistle
+            if (iterator.Owner == owner)
+                continue;
+
+            ExclamateTarget(iterator, component);
```

Comment:
TryComp not Resolve you're going to re-use it across entities and trigger debug asserts.